sklearn/utils/validation.py

```diff
@@ -518,68 +518,166 @@ def _ensure_sparse_format(
+                "When providing 'accept_sparse' as a tuple or list, it must contain at "
                 "least one string value."
             )
         # ensure correct sparse format
-        if spmatrix.format not in accept_sparse:
+        if sparse_container.format not in accept_sparse:
             # create new with correct sparse
-            spmatrix = spmatrix.asformat(accept_sparse[0])
+            sparse_container = sparse_container.asformat(accept_sparse[0])
             changed_format = True
     elif accept_sparse is not True:
         # any other type
         raise ValueError(
-            "Parameter 'accept_sparse' should be a string, "
-            "boolean or list of strings. You provided "
-            "'accept_sparse={}'.".format(accept_sparse)
+            "Parameter 'accept_sparse' should be a string, boolean or list of strings."
+            f" You provided 'accept_sparse={accept_sparse}'."
         )
 
-    if dtype != spmatrix.dtype:
+    if dtype != sparse_container.dtype:
         # convert dtype
-        spmatrix = spmatrix.astype(dtype)
+        sparse_container = sparse_container.astype(dtype)
     elif copy and not changed_format:
         # force copy
-        spmatrix = spmatrix.copy()
+        sparse_container = sparse_container.copy()
 
     if force_all_finite:
-        if not hasattr(spmatrix, "data"):
+        if not hasattr(sparse_container, "data"):
             warnings.warn(
-                "Can't check %s sparse matrix for nan or inf." % spmatrix.format,
+                f"Can't check {sparse_container.format} sparse matrix for nan or inf.",
                 stacklevel=2,
             )
         else:
             _assert_all_finite(
-                spmatrix.data,
+                sparse_container.data,
                 allow_nan=force_all_finite == "allow-nan",
                 estimator_name=estimator_name,
                 input_name=input_name,
             )
 
-    return spmatrix
+    # With SciPy sparse arrays, conversion from DIA format to COO, CSR, or BSR triggers
+    # the use of `np.int64` indices even if the data is such that it could be more
+    # efficiently represented with `np.int32` indices.
+    # https://github.com/scipy/scipy/issues/19245
+    # Since not all scikit-learn algorithms support large indices, the following code
+    # downcasts to `np.int32` indices when it's safe to do so.
+    if (
+        sparse_container_type_name == "dia_array"
+        and changed_format
+        and accept_sparse[0] in ("csr", "coo")
+    ):
+        if accept_sparse[0] == "csr":
+            index_dtype = _smallest_admissible_index_dtype(
+                arrays=(sparse_container.indptr, sparse_container.indices),
+                maxval=max(sparse_container.nnz, sparse_container.shape[1]),
+                check_contents=True,
+            )
+            sparse_container.indices = sparse_container.indices.astype(
+                index_dtype, copy=False
```

Comment:
In case scipy changes the behavior in a future version, this line might be helpful to avoid an extra copy, but this code would become useless anyway.

No strong opinion to keep `copy=False` or not. I am fine with either.

Reply:
> how do we convert an int64 array to an int32 array w/o a copy? O_o

`copy=False` just mean that if you already have 32-bits indices, don't make a copy.